[ffmpeg] avfilter: fix dependencies and linux library ordering

mcmtroffaes opened the pull request:
**Describe the pull request**

The libavfilter component of ffmpeg uses avcodec and avformat, even though ffmpeg itself does not list these as a hard dependency (see https://github.com/FFmpeg/FFmpeg/blob/a16bcc13d9c9db484717c3c19980532fad11c2de/configure#L3705 which lists only avutil). More precisely, it is possible to build ffmpeg with libavfilter but without libavcodec and libavformat, however, the resulting libavfilter library cannot be linked against if libavcodec and libavformat are missing. This patch fixes that problem by listing avcodec and avformat as dependencies of the avfilter feature in vcpkg, similar to how other feature dependencies in ffmpeg are handled.

In addition, the FindFFMPEG cmake script in vcpkg is set up in a way that this library is listed after these components. This breaks linux builds that use libavfilter, with symbols not found errors in the linking stage (see for instance https://travis-ci.com/github/mcmtroffaes/ffmpeg-msvc-build/jobs/479409872#L516). This pull request changes the order in which the libraries are listed, thereby fixing linux builds that rely on libavfilter code. The library ordering rule on linux is that dependencies must follow dependents, so libavfilter must be listed before libavformat and libavcodec.

- What does your PR fix? See above.

- Which triplets are supported/not supported? Have you updated the CI baseline? No changes.

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)? Yes, to the best of my knowledge.

mcmtroffaes committed:
[ffmpeg] avfilter: fix library ordering on linux and fix feature dependencies

mcmtroffaes committed:
[ffmpeg] bump port version

mcmtroffaes commented:
Just found a similar issue with dependencies on OSX being listed in the wrong order. This appears to mainly affect libavfilter as well. Will add another commit to fix this too as part of this PR once my regression tests confirm the fix.

mcmtroffaes committed:
[ffmpeg] add OpenGL, Foundation, CoreGraphics, AppKit and CoreImage for osx (needed for avfilter)

mcmtroffaes committed:
[ffmpeg] update git-tree

mcmtroffaes commented:
Running regression tests with avfilter revealed quite a lot of dependencies for osx to be missing. I don't have a mac myself to test but my osx travis build flagged it:

https://travis-ci.com/github/mcmtroffaes/ffmpeg-msvc-build/jobs/479192010#L393

I've now added these in as well, by looking them up in the apple dev docs at https://developer.apple.com/documentation/. Note these frameworks are listed also in the ffmpeg configure script, so I'm pretty confident this should be all correct now:

* OpenGL: https://github.com/FFmpeg/FFmpeg/blob/2367affc2cea32f18a4d309835bed57adb1baa94/configure#L3536
* Foundation: https://github.com/FFmpeg/FFmpeg/blob/2367affc2cea32f18a4d309835bed57adb1baa94/configure#L3399
* CoreGraphics: https://github.com/FFmpeg/FFmpeg/blob/2367affc2cea32f18a4d309835bed57adb1baa94/configure#L6207
* AppKit: https://github.com/FFmpeg/FFmpeg/blob/2367affc2cea32f18a4d309835bed57adb1baa94/configure#L6194
* CoreImage: https://github.com/FFmpeg/FFmpeg/blob/2367affc2cea32f18a4d309835bed57adb1baa94/configure#L6197

Ready for review now.

mcmtroffaes commented:
... for the record, link to successful build: https://travis-ci.com/github/mcmtroffaes/ffmpeg-msvc-build/jobs/479839991#L313

JackBoosY commented:
cc @cenit for review this PR.

cenit commented:
Those additional dependencies on apple are triggered by the libavfilter, am I right? Because the base port is working also without those. 
All of them are installed with a base dev env, but still I'd think about requiring them only if the libavfilter feature is enabled. What do you think?

mcmtroffaes commented:
Yes, I suspect a lot of these extra osx dependencies could be restricted to specific features. However, I feel somewhat uneasy to make hard assumptions as to what osx libraries are linked to what features without thorough testing. I don't have a mac myself and rely on limited travis testing. For example, the osx OpenGL framework is needed if the avfilter feature is active, but also if OpenGL rendering feature is enabled even if avfilter is not (OpenGL rendering is currently not exposed by vcpkg, but it could be at some point). The CoreImage framework is required by avfilter but nothing else. CoreGraphics and appears to be a dependency for CoreImage.

This said, judging from the link errors, all of these are required for avfilter, and not for anything else in my limited tests, so I'm happy to restrict it as such. Someone who is more vested in macos can still provide further improvements.

mcmtroffaes committed:
[ffmpeg] restrict new osx dependencies to avfilter feature

mcmtroffaes committed:
[ffmpeg] update git-tree

mcmtroffaes commented:
I have patch for it now. Will add the commits here when all regression tests turn green. 🤞

cenit commented:
#15787 is setting up a nice test bench for ffmpeg also here on vcpkg, so you might be interested in waiting for its merge

mcmtroffaes commented:
Thanks for pointing that out. And yes, it would be very nice to have at least part of these tests as part of the vcpkg ci infrastructure. For the record, you can find my own test code here, which aims to allow testing of any arbitrary combination of enabled features: https://github.com/mcmtroffaes/ffmpeg-msvc-build/tree/master/test It's not complete but it covers a good range. Especially for ffmpeg, it is very useful to compile some small sample applications against ffmpeg as part of the testing, since it's the only way to discover issues with the FindFFMPEG script.

mcmtroffaes commented:
I've done the suggested change, @cenit. This is ready again for review.

JackBoosY commented:
Will test features.

JackBoosY commented:
Tested feature avfilter successfully on `x86-windows` and `x64-windows-static`.

Sibras commented:
This is slightly incorrect. Just to be clear avfilter does not depend on avcodec or avformat as a hard requirement. But several features inside avfilter use components of avcodec/avformat if they are available. If avcodec/avformat is disabled then then those filters that require them inside avfilter are also disabled and there is no dependency ([](https://github.com/FFmpeg/FFmpeg/blob/a16bcc13d9c9db484717c3c19980532fad11c2de/configure#L3507)).

The issue your seeing is that by default vcpkg builds all libs and thus builds an avfilter with the avcodec/avformat dependencies that require additionally linking those libs. Perhaps the features in vcpkg should be changed to disable feature instead of enable as they are enabled by default anyway.

Sibras commented on ports/ffmpeg/CONTROL:
```diff
@@ -30,6 +30,7 @@ Build-Depends: ffmpeg[core,avcodec,avformat]
 Description: Device support in ffmpeg
 
 Feature: avfilter
+Build-Depends: ffmpeg[core,avcodec,avformat]
```
As I mentioned earlier, this is incorrect as avfilter only depends on avcodec/avformat if they are enabled. I fthey are not enabled then there is no dependency

mcmtroffaes replied:
This build error says otherwise (this is from an avfilter build without avcodec/avformat): https://ci.appveyor.com/project/mcmtroffaes/ffmpeg-msvc-build/builds/37549914/job/5vckrkqa2jl3064s#L384

``avio_write`` is part of avformat: https://ffmpeg.org/doxygen/trunk/avio_8h.html

How do you suggest fixing this, other than adding avformat (and thereby, avcodec) as a dependency?

mcmtroffaes replied:
I digged a bit deeper. At least for some filters, the ffmpeg configure script does not disable them if they depend on avcodec or avformat. For example, f_metadata definitely depends on avformat, yet this dependency is nowhere listed in the configure script, and it is definitely included in a plain ffmpeg[core,avfilter] build. For example, on linux:

```
./vcpkg install ffmpeg[core,avfilter]
$ cat buildtrees/ffmpeg/src/n4.3.1-3a99257db6.clean/libavfilter/f_metadata.c | grep avformat
#include "libavformat/avio.h"
$ cat buildtrees/ffmpeg/build-x64-linux-rel-out.log | grep f_metadata
CC      libavfilter/f_metadata.o
$ cat buildtrees/ffmpeg/build-x64-linux-rel-out.log | grep avformat
$
```

And indeed, trying to compile any code that depends on just libavformat currently fails to build due to a missing dependency. I've come to conclude that we're not meant to compile avfilter without avcodec/avformat. If so, adding this dependency to vcpkg is the right thing to do. Otherwise, the user ends up with a broken library.

I'd love to have a better solution for this, though.

Sibras replied:
hmm, thats actually a bug in the ffmpeg source itself that would need to be fixed upstream as a build without avcodec/avformat should be able to work without needing avcodec/avformat (clearly though noone else appears to use ffmpeg without codecs in order to run into this bug).

mcmtroffaes replied:
Ok. I've removed these dependencies in favour of fixing the configure script at a later point (will do that as part of another pull request).

Sibras commented on ports/ffmpeg/FindFFMPEG.cmake.in:
```diff
@@ -260,15 +260,38 @@ endif()
+    if (NOT CG_UNIT)
+      message(FATAL_ERROR "CoreGraphics not found")
+    endif()
+    find_library(FND_UNIT Foundation)
```
The avfoundation framework is only used by libavdevice and not avfilter. see your own link for reasoning [https://github.com/FFmpeg/FFmpeg/blob/2367affc2cea32f18a4d309835bed57adb1baa94/configure#L3399](https://github.com/FFmpeg/FFmpeg/blob/2367affc2cea32f18a4d309835bed57adb1baa94/configure#L3399) where _indev means its an input device for libavdevice

mcmtroffaes replied:
Fixed in the latest version as per discussion below.

Sibras commented on ports/ffmpeg/FindFFMPEG.cmake.in:
```diff
@@ -260,15 +260,38 @@ endif()
+    if (NOT CI_UNIT)
+      message(FATAL_ERROR "CoreImage not found")
+    endif()
+    find_library(CG_UNIT CoreGraphics)
```
As below, CoreGraphics is pulled in by avfoundation which is only for avdevice

mcmtroffaes replied:
Fixed in the latest version as per discussion below.

Sibras reviewed (changes requested):
See my previous comment

Sibras commented:
For reference I went through the ffmpeg configure file and isolated the following framework dependencies:

- OpenGL framework is used by the coreimage+coreimagesrc filters in libavfilter
- OpenGL framework will in future (when I add the feature) be used by opengl_out device in libavdevice
- Foundation/AVFoundation framework is used by avfoundation_in in libavdevice
- CoreGraphics is used by avfoundation_in in libavdevice
- Appkit is used by coreimage+coreimagesrc filters in libavfilter
- CoreImage is used by coreimage+coreimagesrc filters in libavfilter
- AudioToolbox is used by a bunch of decoders in libavcodec
- AudioToolbox is used by audiotoolbox_in in libavdevice
- VideoToolbox is used in hwaccel libavutil (also pulls in QuartzCore framework)
- VideoToolbox is used for various decodes in libavcodec
- CoreServices is used by VideoToolbox aka libavutil + libavcodec
- CoreMedia is used by VideoToolbox aka libavutil + libavcodec
- CoreMedia is used in avfoundation in libavdevice
- CoreVideo is used by VideoToolbox aka libavutil + libavcodec
- CoreVideo is used in avfoundation in libavdevice
- OpenCL framework is used by various filters in libavfilter (when opencl options is enabled)
- OpenCL framework is used by hwaccel in libavutil (when opencl options is enabled)
- CoreFoundation by tls_protocol in libavdevice
- Security by tls_protocol in libavdevice

mcmtroffaes commented:
Thanks for the feedback and for the research, @Sibras. I think I forewarned on the limitations of the patch being non-ideal with respect to the macos improvements, being only tested against a standard build (including avcodec, avformat, avdevice, and avfilter). Sadly I cannot properly test mac, and have to go by limited tests on travis. I 'll try to update the patch to reflect your findings, under the understanding that someone else will have to test the individual features.

mcmtroffaes committed:
[ffmpeg] improve macos framework dependencies

mcmtroffaes committed:
[ffmpeg] remove avfilter dependencies

Note this exposes a bug in upstream, will need fix in ffmpeg configure script itself.

mcmtroffaes committed:
[ffmpeg] update git-tree

mcmtroffaes committed:
[ffmpeg] patch for filter dependencies in configure script

mcmtroffaes committed:
[ffmpeg] update git-tree

mcmtroffaes commented:
Well, this was easier than I thought. I have a patch for the configure issue (see https://github.com/mcmtroffaes/vcpkg/blob/b6e300355090c31cb615695ca743a28dcde96357/ports/ffmpeg/0014-avfilter-dependency-fix.patch); now running the regression tests to see if it catches everything.

mcmtroffaes committed:
[ffmpeg] apple framework fix: securetransport is part of avformat and not avdevice

mcmtroffaes committed:
[ffmpeg] update git-tree

Sibras commented on ports/ffmpeg/0014-avfilter-dependency-fix.patch:
```diff
@@ -0,0 +1,23 @@
+ xfade_opencl_filter_deps="opencl"
+ yadif_cuda_filter_deps="ffnvcodec"
+ yadif_cuda_filter_deps_any="cuda_nvcc cuda_llvm"
++ametadata_filter_deps="avformat"
```
Rearrange these so that they are in alphabetical order instead of added to the end. That way you should be able to send this as a patch to ffmpeg and get it upstreamed.

mcmtroffaes replied:
The vcpkg maintainer guidelines are very explicit about the fact that patches should be minimized (logically so, to minimize the chance of the patch breaking with future updates): https://vcpkg.readthedocs.io/en/latest/maintainers/maintainer-guide/ so what you suggest cannot be shipped with vcpkg. However, for upstream, of course.

mcmtroffaes replied:
http://ffmpeg.org/pipermail/ffmpeg-devel/2021-February/275819.html

mcmtroffaes commented:
This should be ready now, all comments have been addressed and my own the regression tests have passed.

ras0219-msft commented:
LGTM, thanks everyone for the awesome work here!

mcmtroffaes commented:
Thank you for merging. Also thanks to @cenit @Sibras for the excellent constructive feedback!
